skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp

```diff
@@ -665,3 +732,1297 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
+  });
+
+  return RE::BSEventNotifyControl::kContinue;
+}
+
+RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
+  const RE::ItemsPickpocketed::Event* event,
+  RE::BSTEventSource<RE::ItemsPickpocketed::Event>* eventSource)
+{
+  if (event == nullptr) {
+    return RE::BSEventNotifyControl::kContinue;
+  }
+
+  auto itemCount = event->numItems;
+
+  SkyrimPlatform::GetSingleton().AddUpdateTask([itemCount] {
+    auto obj = JsValue::Object();
+
+    if (itemCount < 0) {
+      return;
+    }
+
+    obj.SetProperty("count", JsValue::Double(itemCount));
+
+    EventsApi::SendEvent("itemPickpocketed", { JsValue::Undefined(), obj });
+  });
+
+  return RE::BSEventNotifyControl::kContinue;
+}
+
+RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
+  const RE::TESDestructionStageChangedEvent* event,
+  RE::BSTEventSource<RE::TESDestructionStageChangedEvent>* eventSource)
+{
+  auto converted =
+    reinterpret_cast<const TESEvents::TESDestructionStageChangedEvent*>(event);
+
+  if (!converted) {
+    return RE::BSEventNotifyControl::kContinue;
+  }
+
+  auto target = converted->target.get();
+  auto targetId = target ? target->formID : 0;
+  auto oldStage = converted->oldStage;
+  auto newStage = converted->newStage;
+
+  SkyrimPlatform::GetSingleton().AddUpdateTask(
+    [target, targetId, oldStage, newStage] {
+      auto obj = JsValue::Object();
+
+      auto targetLocal = RE::TESForm::LookupByID(targetId);
+
+      if (targetLocal == nullptr || targetLocal != target) {
+        return;
+      }
+
+      obj.SetProperty("target", CreateObject("ObjectReference", targetLocal));
+      obj.SetProperty("oldStage", JsValue::Double(oldStage));
+      obj.SetProperty("newStage", JsValue::Double(newStage));
+
+      EventsApi::SendEvent("destructionStageChanged",
+                           { JsValue::Undefined(), obj });
+    });
+
+  return RE::BSEventNotifyControl::kContinue;
+}
+
+RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
+  const RE::TESSceneActionEvent* event,
+  RE::BSTEventSource<RE::TESSceneActionEvent>* eventSource)
+{
+  auto converted =
+    reinterpret_cast<const TESEvents::TESSceneActionEvent*>(event);
+
+  if (!converted) {
+    return RE::BSEventNotifyControl::kContinue;
+  }
+
+  auto ref = converted->reference;
+  auto refId = ref ? ref->formID : 0;
+  auto sceneId = converted->sceneId;
+  auto refAliasId = converted->referenceAliasID;
+  auto questId = converted->questId;
+  auto action = converted->action;
+
+  SkyrimPlatform::GetSingleton().AddUpdateTask(
+    [ref, refId, sceneId, refAliasId, questId, action] {
+      auto obj = JsValue::Object();
+
+      auto refLocal = RE::TESForm::LookupByID(refId);
+      auto sceneLocal = RE::TESForm::LookupByID(sceneId);
+      auto questLocal = RE::TESForm::LookupByID(questId);
+
+      if (refLocal == nullptr || sceneLocal == nullptr ||
+          questLocal == nullptr || refAliasId < 0 || action < 0 ||
+          sceneLocal->formType != RE::FormType::Scene ||
```

Comment:
check equality to original values instead of checking form type.

Reply:
I should check `sceneLocal->formID == sceneId` after `auto sceneLocal = RE::TESForm::LookupByID(sceneId)` ?

Reply:
check `refLocal == ref` never pass, it is probably not form